### PR TITLE
[MRG + 1] stac-oneoff weave (situated communication)

### PR DESCRIPTION
**Pull request by kowey**
Utility to merge an augmented corpus into a pre-existing one.
It may be a bit buggy, but we don't have to use it immediately.

This could be an MRG if we want

### CODE REVIEW

**Comment by kowey** [comment]
Note that _hollow_out_nonplayer_text (from educe.stac.oneoff.cmd.weave) sounds like it it duplicates the functionality of that turn star blanking out function.  At least from the name of the function

**Comment by moreymat** [comment]
I have read the code but need to re-review this tomorrow with a fresher brain.

File: educe/stac/oneoff/cmd/weave.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    """Short text representation of an annotation"""
+    span = anno.text_span()
+    text = doc.text(span)
+    if len(text) > max_width:
+        snippet = text[:max_width] + '...'
+    else:
+        snippet = text
+    template = '{ty} {span} [{snippet}]'
+    return template.format(ty=anno.type,
+                           span=anno.text_span(),
```
Comment (moreymat) [inline educe/stac/oneoff/cmd/weave.py]: Maybe you could reuse the `span` variable defined above as `anno.text_span()`.

File: educe/stac/oneoff/cmd/weave.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+
+def _hollow_out_nonplayer_text(src_doc):
+    """Return a version of the source text where all characters in nonplayer
+    turns are replaced with a nonsense char (tab).
+
+    This is to prevent spurious matching by the sequence matcher.
+    """
+    # we can't use the API one until we update it to account for the
+    # fancy new identifiers
```
Comment (moreymat) [inline educe/stac/oneoff/cmd/weave.py]: Not sure I get this comment.

File: educe/stac/oneoff/cmd/weave.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    non_player_spans = [x.text_span() for x in src_doc.units
+                        if x.type == 'NonplayerTurn']
+    current = None
+    merged = []
```
Comment (moreymat) [inline educe/stac/oneoff/cmd/weave.py]: Maybe a less generic name like `merged_np_spans` would help to understand this code faster :-)

File: educe/stac/oneoff/weave.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+    # case 2 and 5 (to be pruned below)
+    res.expected_src_only.extend(src_doc.units)
+    res.abnormal_tgt_only.extend(tgt_doc.units)
+
+    # case 1, 2 and 4
+    for src, tgt, size in matches:
+        tgt_to_src = src - tgt
+        res.shift_if_ge[tgt] = tgt_to_src  # case 1 and 2
+        src_annos = _units_between(src_doc, src, src + size)
```
Comment (moreymat) [inline educe/stac/oneoff/weave.py]: Could you unroll `_units_between` in these two lines and kill the said function?
It is a simple one-liner that calls another simple one-liner (`educe.stac.context.enclosed`) that itself calls what could be (written as) a simple one-liner (`educe.annotation.Span.encloses`). This is usually a bad sign.

File: educe/stac/oneoff/weave.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+                                [fun(x) for x in self.abnormal_src_only],
+                                [fun(x) for x in self.abnormal_tgt_only],
+                                [fun(x) for x in self.expected_src_only])
+
+# ---------------------------------------------------------------------
+#
+# ---------------------------------------------------------------------
+
+
+def src_gaps(matches):
```
Comment (moreymat) [inline educe/stac/oneoff/weave.py]: I cannot find any call to this function.

File: educe/stac/oneoff/cmd/weave.py (hunk 1 above)
Comment (kowey) [inline educe/stac/oneoff/cmd/weave.py]: k

File: educe/stac/oneoff/cmd/weave.py (hunk 2 above)
Comment (kowey) [inline educe/stac/oneoff/cmd/weave.py]: Will try to flesh this out a bit more

File: educe/stac/oneoff/cmd/weave.py (hunk 3 above)
Comment (kowey) [inline educe/stac/oneoff/cmd/weave.py]: k

File: educe/stac/oneoff/weave.py (hunk 4 above)
Comment (kowey) [inline educe/stac/oneoff/weave.py]: Eh, ok.  I guess I was going for something semantic'ish, but it's not clear to me that my units_between is any clearer than the one liner it replaces, so fine.  It's not about golf.

File: educe/stac/oneoff/weave.py (hunk 5 above)
Comment (kowey) [inline educe/stac/oneoff/weave.py]: I'll let you kill this one after one of us has had a chance to debug this code.  Right now, I'm going to leave it in for symmetry.

**Comment by moreymat** [comment]
Is this working ok or still kinda broken ? (cf. catchup/situated)

**Comment by kowey** [comment]
It unfortunately still crashes on the one example I tried.
I should try it on another example to see if it's just that one

**Pull request merged**